src/ScottPlot5/ScottPlot5/DataSourceUtilities.cs

```diff
@@ -0,0 +1,475 @@
+        {
+            renderInfo.TryGetPixelPerUnitX(xaxis, out var pixelScaleX);
+            renderInfo.TryGetPixelPerUnitY(yaxis, out var pixelScaleY);
+
+            return new Coordinates(
+                x: UnScale(pixelCoordinate.X, pixelScaleX, xScalingFactor, xOffset),
+                y: UnScale(pixelCoordinate.Y, pixelScaleY, yScalingFactor, yOffset)
+                ); ;
+
+            static double UnScale(double value, double pixelScaling, double axisScaling, double offset)
+            {
+                return ((value / pixelScaling) - offset) / axisScaling;
+            }
+        }
+
+        /// <summary>
+        /// Get the nearest datapoint on the plot from the <paramref name="dataSource"/>, based on the <paramref name="mouseLocation"/> and the <paramref name="renderInfo"/>
+        /// </summary>
+        /// <remarks>This is the original way to locate the nearest DataPoint from the collection, and is safe for unsorted collections (such as Scatter)</remarks>
+        /// <param name="dataSource">The data source</param>
+        /// <param name="mouseLocation">the mouse coordinates from the plot. <see cref="Plot.GetCoordinates(Pixel, IXAxis?, IYAxis?)"/></param>
+        /// <param name="renderInfo"><see cref="Plot.LastRender"/></param>
+        /// <param name="maxDistance">The maximum distance to search</param>
+        /// <param name="xAxis">The X-Axis of assigned to the datasource. If not specified, uses the bottom axis.</param>
+        /// <param name="yAxis">The X-Axis of assigned to the datasource. If not specified, uses the left axis.</param>
+        /// <returns>
+        /// If match found : returns a datapoint that represents the closest (X,Y) coordinate on the plot, and the index that can be used to get the values from the DataSource.
+        /// <br/>If no match found : returns <see cref="DataPoint.None"/>
+        /// </returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static DataPoint GetNearest(IDataSource dataSource, Coordinates mouseLocation, RenderDetails renderInfo, float maxDistance, IXAxis? xAxis = null, IYAxis? yAxis = null)
+        {
+            if (dataSource is null) throw new ArgumentNullException(nameof(dataSource));
+
+            var renderIndexCount = dataSource.GetRenderIndexCount();
+            var minRenderIndex = dataSource.MinRenderIndex;
+
+            double maxDistanceSquared = maxDistance * maxDistance;
+            double closestDistanceSquared = double.PositiveInfinity;
+
+            int closestIndex = 0;
+
+            _ = renderInfo.TryGetPixelPerUnitX(xAxis, out double pxPerUnitX);
+            _ = renderInfo.TryGetPixelPerUnitY(yAxis, out double pxPerUnitY);
+
+            bool preferCoordinates = dataSource.PreferCoordinates;
+
+            for (int i2 = 0; i2 < renderIndexCount; i2++)
```

Comment:
Good spot, I'll review. Something I did notice is that most max render index values are NOT SET! I crashed the demo application today adding a scatter plot because it's set to the int max, not set to the `array length - 1`

Reply:
Such a pre-check should solve the problems you describe
```c#
var maxRenderIndex = Math.Min(dataSource.MaxRenderIndex, dataSource.Length - 1);
 ... 
for (int i = minRenderIndex; i <= maxRenderIndex; i++)
```

Reply:
What I'm thinking is review all of the data sources, any sources that use arrays can have this property automatically set during construction instead of being sent positive infinity or whatever. (Lists will still have to have this property set to positive Infinity by default)

For IDataSource.Length I am going to take whatever the minimum of the arrays or lists are. 
I've also already added a helper method to get that max render index using the calculation you described.